### PR TITLE
[JIT] Support a single TensorList argument anywhere in the argument list

### PR DESCRIPTION
**Pull request by jamesr66a**

### CODE REVIEW

File: cmake/Modules/FindMKL.cmake [hunk 1]
```diff
@@ -273,6 +273,7 @@ ELSE (MKL_LIBRARIES)
   SET(MKL_VERSION)
 ENDIF (MKL_LIBRARIES)
 
+set(MKL_FIND_REQUIRED Off)
```
Comment (apaszke) [inline cmake/Modules/FindMKL.cmake]: I don't think you wanted to include this one in this patch

File: tools/jit/gen_jit_dispatch.py [hunk 2]
```diff
@@ -97,9 +97,11 @@ def is_jit_op(decl):
...
     # and the only tensor argument
     arguments = decl['arguments']
-    has_tensorlist = any(arg['simple_type'] == 'TensorList' for arg in arguments)
-    num_tensor_args = sum(map(is_tensor_arg, arguments))
-    if has_tensorlist and (num_tensor_args != 1 or arguments[0]['simple_type'] != 'TensorList'):
+    # Only support a single TensorList arg
+    if sum(arg['simple_type'] == 'TensorList' for arg in arguments) > 1:
+        return False
+    # omit einsum
+    if any(arg['simple_type'] == 'std::string' for arg in arguments):
```
Comment (apaszke) [inline tools/jit/gen_jit_dispatch.py]: Why not add it to `blacklisted_types`?

File: tools/jit/gen_jit_dispatch.py [hunk 3]
```diff
@@ -162,15 +164,27 @@ def emit_decl_variant(decl, is_positional_arg, has_tensorlist):
             num_dynamic_inputs = static_inputs
 
         real_inputs = count()
+        if has_tensorlist:
+            kw_assignments.append('size_t _Arg_idx = 0;')
```
Comment (apaszke) [inline tools/jit/gen_jit_dispatch.py]: `_Arg_idx` doesn't really say what it's for. Can you call it `_tensorlist_idx`?

File: tools/jit/gen_jit_dispatch.py [hunk 4]
```diff
@@ -162,15 +164,27 @@ def emit_decl_variant(decl, is_positional_arg, has_tensorlist):
...
         for i, arg in enumerate(decl['arguments']):
             # XXX: we currently support only TensorList ops that have a TensorList as
             # the first argument, that is then followed by a number of positional args.
             if arg['simple_type'] == 'TensorList':
-                arguments.append('peekSlice(stack, 0, varargs_length - {}, varargs_length)'.format(static_inputs))
+                kw_assignments.append('size_t _Arg_idx_{} = _Arg_idx;'.format(i))
```
Comment (apaszke) [inline tools/jit/gen_jit_dispatch.py]: Wait, what's going on in here 😮 Can you please post an autogenerated snippet?

File: tools/jit/gen_jit_dispatch.py [hunk 5]
```diff
@@ -162,15 +164,27 @@ def emit_decl_variant(decl, is_positional_arg, has_tensorlist):
             num_dynamic_inputs = static_inputs
 
         real_inputs = count()
+        if has_tensorlist:
+            kw_assignments.append('size_t _Arg_idx = 0;')
         for i, arg in enumerate(decl['arguments']):
             # XXX: we currently support only TensorList ops that have a TensorList as
```
Comment (zdevito) [inline tools/jit/gen_jit_dispatch.py]: This comment is out of date.

File: tools/jit/gen_jit_dispatch.py [hunk 6]
```diff
@@ -162,15 +164,27 @@ def emit_decl_variant(decl, is_positional_arg, has_tensorlist):
...
+                kw_assignments.append('size_t _Arg_idx_{} = _Arg_idx;'.format(i))
+                kw_assignments.append('_Arg_idx += varargs_length - {};'.format(static_inputs))
+                arguments.append(
+                    'peekSlice(stack, _Arg_idx_{}, varargs_length - {}, varargs_length)'.format(i, static_inputs))
             elif arg['simple_type'] in default_only_types:
                 arguments.append(arg['default'])
             elif is_tensor_arg(arg):
-                arguments.append('std::move(peek(stack, {}, {}))'.format(next(real_inputs), static_inputs))
+                if has_tensorlist:
+                    kw_assignments.append('size_t _Arg_idx_{} = _Arg_idx++;'.format(i))
```
Comment (zdevito) [inline tools/jit/gen_jit_dispatch.py]: I am pretty sure there are missing cases for the `is_positional_arg` cases below. Changing all of the sites that use `peek` in this way is pretty fragile.

**Review by zdevito — COMMENTED** [review]
I think a tweak to this approach would make it more robust:

Let's say you have
```
a b c [d e f g] h i # [] is the list
```

We extract  a, b, c, with `peek(0, N)` where N is the number of arguments. We can adjust N before and after the varargs list is encountered to make `peek(i, N)`  still always extract the `i`th static argument. 

Before tensor_list, set N to  `(num_static_args - 1 + num_args_in_list)`
```
a b c [d e f g] h i # [] is the list
~~~~~~~~~~~~ <- N
```

When we handle the varargs_list, then adjust N to be 
`num_static_args`

```
a b c [d e f g] h i # [] is the list
              ~~~~~~ <- N
```

**Comment by jamesr66a** [comment]
Superseded by https://github.com/pytorch/pytorch/compare/master...jamesr66a:tensorlist2?expand=1

**Pull request closed without merging**